python/lsst/ctrl/mpexec/execFixupDataId.py

```diff
@@ -0,0 +1,103 @@
+# This file is part of ctrl_mpexec.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ['ExecutionGraphFixup']
+
+from typing import Any, Iterable, Tuple, Union
+
+from lsst.pipe.base import QuantumIterData
+from .executionGraphFixup import ExecutionGraphFixup
+
+
+class ExecFixupDataId(ExecutionGraphFixup):
+    """Implementation of ExecutionGraphFixup for ordering of tasks based
+    on DataId values.
+
+    This class is a trivial implementation mostly useful as an example,
+    though it can be used to make actual fixup instances by defining
+    a method that instantiates it, e.g.::
+
+        # lsst/ap/verify/ci_fixup.py
+
+        from lsst.ctrl.mpexec.execFixupDataId import ExecFixupDataId
+
+        def assoc_fixup():
+            return ExecFixupDataId(taskLabel="ap_assoc",
+                                   dimensions=("visit", "detector"))
+
+
+    and then executing pipetask::
+
+        pipetask run --graph-fixup=lsst.ap.verify.ci_fixup.assoc_fixup ...
+
+    This will add new dependencies between quanta executed by the task with
+    label "ap_assoc". Quanta with higher visit number will depend on quanta
+    with lower visit number and their execution will wait until lower visit
+    number finishes.
+
+    Parameters
+    ----------
+    taskLabel : `str`
+        The label of the task for which to add dependencies.
+    dimensions : `str` or iterable of `str`
```

Comment:
If I understand the code right, this should be "sequence of `str`". The sort order would be ambiguous if you passed in, for example, a `set`.

Reply:
Preferred syntax [uses bracket notation](https://developer.lsst.io/python/numpydoc.html#py-docstring-parameter-types-sequences).

Reply:
Indeed, the code as it is written now won't even work with iterable. I'm inclined to change the arguments to positionals:

    def __init__(self, taskLabel: str, *dimensions: str, *, reverse: bool = False)

to avoid that ordering ambiguity. Do you like this better?

Regarding bracket notation - I'm not quite sure that it applies here, strictly speaking iterable is not a container, and I did not find an example in the developer guide for that, that is why I decided to use `iterable of`

Reply:
I don't understand why you would want to use variadic arguments here, they seem more error-prone. Why can't you simply document the argument as a sequence rather than any iterable?

Reply:
I like variadic better because it fixes the order and most of the time dimensions will be known in advance so in my view it's more natural to write:

    Class("label", "visit", "detector")

than 

    Class("label", ("visit", "detector"))

Still if you prefer single argument I'll do it that way.